### PR TITLE
Add ability to export EQ items as QTI 3.0

### PR DESCRIPTION
**Pull request by wtritch**
This PR exports EQ items as QTI 3.0.

Things to note:
1. I'm using the TIMS EqTransformer (with the required TransformerFactory, ImageGenerator, and StringAssembler references) to transform the JSON into EAX version 2 (EQ input definition XML.)
2. Since the EQ Scoring tab introduces a system-defined Variable-per-Part, I needed to break the existing  QtiScoringMapper interface/implementation into a higher-fidelity API.  It was previously processing the QtiResponseProcessing model instance in it's entirety, but with the system-declared/undefined ResponseN variables, consumers need to be able to inject additional response processing operators.

### CODE REVIEW

File: item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+import static org.opentestsystem.ap.item.translation.mappers.qti.CommonQtiAssessmentItemMapper.OrderCommonQtiMapper;
+import static org.opentestsystem.ap.item.translation.mappers.qti.QtiConstants.NamespaceQti;
+import static org.opentestsystem.ap.item.translation.mappers.qti.QtiConstants.ResponseIdentifier;
+
+/**
+ * This mapper is responsible for mapping EQ-specific content and attributes
+ * to the given QTI assessment item.
+ */
+@Component
+class EqQtiMapper implements QtiAssessmentItemMapper {
```
Comment (wtritch) [inline item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java]: Here's the new EQ-focused item JSON -> QTI mapper.  All other changes are in support of this.

File: item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/QtiScoringMapper.java [hunk 2]
```diff
@@ -19,16 +24,57 @@
...
+
+    /**
+     * Given a saturated Jsoup QTI response processing element, convert the given
+     * DOM into a QtiResponseProcessing model instance.
+     *
+     * @param responseProcessingElement A qti response processing element
+     * @return  The associated QtiResponseProcessing instance
      */
-    QtiResponseProcessing mapResponseProcessing(final String itemType, final Scoring scoring);
+    QtiResponseProcessing transformToQtiResponseProcessing(Element responseProcessingElement);
```
Comment (wtritch) [inline item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/QtiScoringMapper.java]: These are the scoring mapper changes mentioned in the overview:
`QtiResponseProcessing mapResponseProcessing(final String itemType, final Scoring scoring)`
->
`Element createResponseProcessingElement()`
`Element mapToPreProcessResponseOutcomeVariable(Document document,
                                                   String itemType,
                                                   Variable ppResponseVariable)`
`Element mapUserVariableToSetOutcomeValue(Document document, Variable variable)`
`Element mapToResponseCondition(Document document, Scoring scoring)`
`QtiResponseProcessing transformToQtiResponseProcessing(Element responseProcessingElement)`

File: item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/GiQtiMapper.java [hunk 3]
```diff
@@ -29,7 +32,8 @@
 import static org.opentestsystem.ap.item.translation.mappers.qti.QtiConstants.ResponseIdentifier;
 
 /**
- * This mapper is responsible for mapping GI-specific content and attributes to the given QTI assessment item.
+ * This mapper is responsible for mapping GI-specific content and attributes
+ * to the given QTI assessment item.
  */
 @Component
 class GiQtiMapper implements QtiAssessmentItemMapper {
```
Comment (wtritch) [inline item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/GiQtiMapper.java]: The existing GI mapper had to be changed to account for the QtiScoringMapper API changes.  Thankfully, the integration tests validate that this is functionally equivalent.

File: item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+     * array value.
+     *
+     * @param parentDocument                The parent document
+     * @param preprocessResponseVariable    The preprocessed response variable
+     * @param variable                      The variable to define
+     * @return A response-processing command element
+     */
+    private Optional<org.jsoup.nodes.Element> mapPartVariable(final org.jsoup.nodes.Document parentDocument,
+                                                              final Variable preprocessResponseVariable,
+                                                              final Variable variable) {
```
Comment (wtritch) [inline item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java]: This is defining the processing around the TIMS-declared "ResponseN" Part variables.  We're just setting Response1 to the pre-processed response variable's first value, Response2 to the second value, etc.

FYI @agorina I'm not sure we interpreted the TDS_ItemScoring implementation properly.  I wonder if we can only have multiple values in the PREPROCESSRESPONSE variable if the item has multiple Parts (multiple equation input fields.)

File: item-translation-service/src/test/resources/items/eq-183358.json [hunk 5]
```diff
@@ -0,0 +1,346 @@
+{
+  "id": "200001",
```
Comment (wtritch) [inline item-translation-service/src/test/resources/items/eq-183358.json]: I hand-jammed a re-creation of 183358 with some minor changes:
1. I added some MathML and plain-text left/right part labels
2. I added a second Part input for testing
3. I manually converted the EAX into TIMS JSON
3. I manually converted the QRX into TIMS Scoring JSON

File: item-translation-service/src/test/resources/items/eq-183358.qti.xml [hunk 6]
```diff
@@ -0,0 +1,99 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<qti-assessment-item
+        xsi:schemaLocation="http://www.imsglobal.org/xsd/imsqtiasi_v3p0 https://purl.imsglobal.org/spec/qti/v3p0/schema/xsd/imsqti_asiv3p0_v1p0.xsd"
+        xmlns="http://www.imsglobal.org/xsd/imsqtiasi_v3p0"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" identifier="sbac-200-200001" title="200001"
+        tool-name="TIMS QTI 3.0 Export" tool-version="5.0.0-SNAPSHOT" time-dependent="false">
```
Comment (wtritch) [inline item-translation-service/src/test/resources/items/eq-183358.qti.xml]: Here's the associated sample EQ Item as QTI 3.0

File: item-translation-service/src/test/resources/items/eq-183358.qti.xml [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    </tab>
+  </tabConfig><editorRow><math xmlns="http://www.w3.org/1998/Math/MathML" title=""><mstyle><mfrac><mn>   4  </mn><mn>   5  </mn></mfrac><mi>  x </mi><mo>  + </mo><mn>  2 </mn><mo>  = </mo><mi>with\space</mi><mi>text</mi><mi>\space</mi><mo>##</mo><mi>\space</mi><mi>some\space</mi><mi>text</mi></mstyle>
+    </math>
+  </editorRow><editorRow><math xmlns="http://www.w3.org/1998/Math/MathML" title=""><mstyle><mi>another</mi><mi>\space</mi><mo>##</mo></mstyle>
+    </math>
+  </editorRow>
+
+</editorconfig>]]>
+                    </template>
```
Comment (wtritch) [inline item-translation-service/src/test/resources/items/eq-183358.qti.xml]: Above is the EAX spit out by the TIMS EqTransformer, just shoved into a \<template> element within the EQ custom interaction.

File: item-translation-service/src/test/resources/items/eq-183358.qti.xml [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            </qti-response-if>
+            <qti-response-else>
+                <qti-set-outcome-value identifier="SCORE">
+                    <qti-base-value base-type="float">
+                        0.0
+                    </qti-base-value>
+                </qti-set-outcome-value>
+            </qti-response-else>
+        </qti-response-condition>
+    </qti-response-processing>
```
Comment (wtritch) [inline item-translation-service/src/test/resources/items/eq-183358.qti.xml]: This response processing element is the result of QRX -> TIMS JSON -> QTI 3.0

File: item-translation-service/src/test/resources/items/eq-183358.qti.xml [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                <qti-variable identifier="CorrectEquation"/>
+            </qti-container-size>
+        </qti-set-outcome-value>
+        <qti-response-condition>
+            <qti-response-if>
+                <qti-equal>
+                    <qti-variable identifier="CorrectEquationCount"/>
+                    <qti-base-value base-type="float">
+                        1
+                    </qti-base-value>
```
Comment (wtritch) [inline item-translation-service/src/test/resources/items/eq-183358.qti.xml]: The QRX called out comparing a "float" to an "integer" ... I hope that works, or automating the QRX -> TIMS JSON may get a little painful.  @agorina ?

File: item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/DefaultQtiScoringMapper.java [hunk 10]
```diff
@@ -122,15 +98,17 @@ private Element mapToPreProcessResponseOutcomeVariable(final Document document,
         return setValueOperator;
     }
 
-    private Element mapVariableToSetOutcomeValue(final Document document, final Variable variable) {
+    @Override
+    public Element mapUserVariableToSetOutcomeValue(final Document document, final Variable variable) {
```
Comment (millerts) [inline item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/DefaultQtiScoringMapper.java]: In the interface you noted that this should only be called when `Variable#isUserVariable()} is True`.  Should you add a conditional and throw if this is called and that isn't the case?  Or return without doing anything if it isn't true?

Note I don't necessarily follow all the design and how QTI works so the above may be a stupid suggestion.

File: item-translation-service/src/test/resources/items/eq-183358.qti.xml (hunk 9 above)
Comment (agorina) [reply item-translation-service/src/test/resources/items/eq-183358.qti.xml]: This works now, not a problem.

File: item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java (hunk 4 above)
Comment (agorina) [reply item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java]: I am not sure I understand the question right, sorry. 

Here is how the response for a multi-part EQ question looks:
```
 "    <math xmlns=\"http://www.w3.org/1998/Math/MathML\" title=\"8+8+8=24-2=22\">\n" +
            "        <mstyle>\n" +
            "            <mn>8</mn>\n" +
            "            <mo>+</mo>\n" +
            "            <mn>8</mn>\n" +
            "            <mo>+</mo>\n" +
            "            <mn>8</mn>\n" +
            "            <mo>=</mo>\n" +
            "            <mn>24</mn>\n" +
            "            <mo>-</mo>\n" +
            "            <mn>2</mn>\n" +
            "            <mo>=</mo>\n" +
            "            <mn>22</mn>\n" +
            "        </mstyle>\n" +
            "    </math>\n" +
            "    <math xmlns=\"http://www.w3.org/1998/Math/MathML\" title=\"s=36\">\n" +
            "        <mstyle>\n" +
            "            <mi>s</mi>\n" +
            "            <mo>=</mo>\n" +
            "            <mn>36</mn>\n" +
            "        </mstyle>\n" +
            "    </math>\n" +
            "</response>\n");
```

Does it help? If not, let's chat.

File: item-translation-service/src/test/resources/items/eq-183358.qti.xml [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        </qti-set-outcome-value>
+        <qti-set-outcome-value identifier="Response1">
+            <qti-index n="1">
+                <qti-variable identifier="Responses"/>
+            </qti-index>
+        </qti-set-outcome-value>
+        <qti-set-outcome-value identifier="Response2">
+            <qti-index n="2">
+                <qti-variable identifier="Responses"/>
+            </qti-index>
```
Comment (agorina) [inline item-translation-service/src/test/resources/items/eq-183358.qti.xml]: I know it is a hand-jammed item, but I do not think the scoring rubric ever relies on the order of the responses. This tells me that the order is not reliable, and I have seen that at least in GI items. It is OK for this test, but a little misleading.

File: item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+        target.setQtiResponseProcessing(mapResponseProcessing(eqItem.getCore().getScoring()));
+
+        final Div contentRoot = contentMapper.appendContentRoot(target);
+        final List content = contentRoot.getContent();
+        content.add(createStem(context, eqItem));
+        content.add(createGiInteraction(context, eqItem));
+
+    }
+
+    private CustomInteractionType createGiInteraction(final QtiTranslationContext<QtiAssessmentItem> context, final EqItem eqItem) {
```
Comment (agorina) [inline item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java]: Looks like the method name is "c&p" that needs an adjustment?

File: item-translation-service/src/test/java/org/opentestsystem/ap/item/translation/service/impl/QtiAssessmentItemTranslationServiceIT.java [hunk 13]
```diff
@@ -58,7 +58,10 @@ public void itShouldGetItemDefinitionsInQti() {
                     loadFileAsString("/items/ebsr-209053.qti.xml")),
             ImmutablePair
                 .of(loadJsonFileAsClass("/items/gi-183417.json", GiItem.class),
-                    loadFileAsString("/items/gi-183417.qti.xml"))
+                    loadFileAsString("/items/gi-183417.qti.xml")),
+            ImmutablePair
+                .of(loadJsonFileAsClass("/items/eq-183358.json", GiItem.class),
+                    loadFileAsString("/items/eq-183358.qti.xml"))
```
Comment (agorina) [inline item-translation-service/src/test/java/org/opentestsystem/ap/item/translation/service/impl/QtiAssessmentItemTranslationServiceIT.java]: Thank you for this! :)

File: item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/DefaultQtiScoringMapper.java (hunk 10 above)
Comment (wtritch) [reply item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/DefaultQtiScoringMapper.java]: Not silly at all.  That's a good idea, although I believe TIMS *currently* always initializes a "Responses" pre-processed response variable in the Scoring tab.

File: item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java (hunk 12 above)
Comment (wtritch) [reply item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java]: Yup, nice catch >_<

File: item-translation-service/src/test/resources/items/eq-183358.qti.xml (hunk 11 above)
Comment (wtritch) [reply item-translation-service/src/test/resources/items/eq-183358.qti.xml]: This is more about accounting for the TIMS UI, which initializes these variables and doesn't let the user remove them:
image
No existing QRX will be referencing them, of course.

File: item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java (hunk 4 above)
Comment (wtritch) [reply item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java]: Yup, and after pre-processing that'll get turned into two values inside PP_RESPONSE? (in existing QRX, PP_RESPONSE is renamed to "Responses" in the TIMS JSON)

File: item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/DefaultQtiScoringMapper.java (hunk 10 above)
Comment (wtritch) [reply item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/DefaultQtiScoringMapper.java]: done

File: item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+
+    private final ObjectFactory objectFactory;
+    private final QtiContentMapper contentMapper;
+    private final CommonQtiMapperSupport commonSupport;
+    private final QtiScoringMapper scoringMapper;
+    private final TransformerFactory transformerFactory;
+    private final ImageGenerator imageGenerator;
+    private final StringAssembler stringAssembler;
+
+    private volatile EqTransformer eqTransformer;
```
Comment (esotrope) [inline item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java]: uh oh. this is never a good sign ; p

File: item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java (hunk 4 above)
Comment (agorina) [reply item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java]: Yes, after pre-processing that'll get turned into two values inside PP_RESPONSE.

A single response like this: 3+x=7-y=3*x/y will be pre-processed into two values inside of PP_RESPONSE too.

Comment (wtritch) [reply item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java]: It’s that second case that concerns me.

Comment (agorina) [reply item-translation-service/src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EqQtiMapper.java]: Could you please clarify your concern? If it just the user can even enter it? Or do you have any other concerns?

**Pull request merged**